File: java/formula-java/src/main/java/org/formula/optimize/FormulaOptimizer.java
package org.formula.optimize;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import org.formula.NamedResolvedValue;
import org.formula.QuotedTextResolvedValue;
import org.formula.ResolvedValue;
import org.formula.parse.shuntingyard.Associativity;
import org.formula.parse.shuntingyard.OperatorFunction1;
import org.formula.parse.shuntingyard.OperatorFunction2;
import org.formula.parse.shuntingyard.OperatorFunction3;
import org.formula.parse.shuntingyard.ShuntingYardParser;
import org.formula.util.Lambda1;
import org.formula.util.Lambda2;
import org.formula.util.Lambda3;

public class FormulaOptimizer {

    private static final ShuntingYardParser PARSER = ShuntingYardParser.create()
            .operator("^", 4, Associativity.RIGHT, opFn2((a, b) -> a + "^" + b))
            .operator("*", 3, Associativity.LEFT, (a,b) -> new MathFunction("*", a, b))
            .operator("/", 3, Associativity.LEFT, (a,b) -> new MathFunction("/", a, b))
            .operator("+", 2, Associativity.LEFT, (a,b) -> new MathFunction("+", a, b))
            .operator("-", 2, Associativity.LEFT, (a,b) -> new MathFunction("-", a, b))
            .operator("!", 2, Associativity.LEFT, opFn1((a) -> "!" + a))
            .operator("<", 3, Associativity.LEFT, opFn2((a, b) -> a + "<" + b))
            .operator("<=", 3, Associativity.LEFT, opFn2((a, b) -> a + "<=" + b))
            .operator(">", 3, Associativity.LEFT, opFn2((a, b) -> a + ">" + b))
            .operator(">=", 3, Associativity.LEFT, opFn2((a, b) -> a + ">=" + b))
            .operator("==", 3, Associativity.LEFT, opFn2((a, b) -> a + "==" + b))
            .operator("!=", 3, Associativity.LEFT, opFn2((a, b) -> a + "!=" + b))
            .operator("AND", 1, Associativity.LEFT, (a, b) -> new AllFunction(List.of(b, a)))
            .operator("OR", 1, Associativity.LEFT, (a, b) -> new AnyFunction(List.of(b, a)))
            .operator("d", 4, Associativity.LEFT, opFn2((a, b) -> a + "d" + b))
            .term("true", () -> ResolvedValue.of("true"))
            .term("false", () -> ResolvedValue.of("false"))
            .term("null", () -> ResolvedValue.of("null"))
            .function("abs", opFn1("abs(%s)"::formatted))
            .function("min", opFn2("min(%s,%s)"::formatted))
            .function("max", opFn2("max(%s,%s)"::formatted))
            .function("floor", opFn1("floor(%s)"::formatted))
            .function("ceil", opFn1("ceil(%s)"::formatted))
            .function("signed", opFn1("signed(%s)"::formatted))
            .function("if", opFn3("if(%s,%s,%s)"::formatted))
            .function("concat", opFn2("concat(%s,%s)"::formatted))
            .function("ordinal", opFn1("ordinal(%s)"::formatted))
            .function("any", AnyFunction::new)
            .function("all", AllFunction::new)
            .variable("@", (context, key) -> ResolvedValue.of("@%s".formatted(key)))
            .variable("@{", "}", (context, key) -> ResolvedValue.of("@{%s}".formatted(key)))
            .variable("min(@", ")", (context, key) -> ResolvedValue.of("min(@%s)".formatted(key)))
            .variable("max(@", ")", (context, key) -> ResolvedValue.of("max(@%s)".formatted(key)))
            .variable("sum(@", ")", (context, key) ->ResolvedValue.of("sum(@%s)".formatted(key)))
            .comment("[", "]", (value, comment) -> NamedResolvedValue.of(value, comment.substring(1, comment.length() - 1), "[", "]"))
            ;

    public static String optimize(String formulaText) {
        var resolved = FormulaOptimizer.PARSER.parse(formulaText).resolve();
        if (resolved instanceof MathFunction mf) {
            return mf.asTextNoBrackets();
        }
        return format(resolved);
    }

    private static String format(ResolvedValue value) {
        if (value instanceof QuotedTextResolvedValue quoted) {
            return quoted.asQuotedText();
        }
        if (value instanceof NamedResolvedValue named) {
            return named.toString();
        }
        return value.asText();
    }

    private static OperatorFunction1 opFn1(Lambda1<String, String> fn) {
        return (ResolvedValue a) -> ResolvedValue.of(fn.execute(format(a)));
    }

    private static OperatorFunction2 opFn2(Lambda2<String, String, String> fn) {
        return (ResolvedValue a, ResolvedValue b) -> ResolvedValue.of(fn.execute(format(a), format(b)));
    }

    private static OperatorFunction3 opFn3(Lambda3<String, String, String, String> fn) {
        return (ResolvedValue a, ResolvedValue b, ResolvedValue c) -> ResolvedValue.of(fn.execute(format(a), format(b), format(c)));
    }

    private static class MathFunction extends AbstractOptimizedFunction {
        private final String operator;
        private final ResolvedValue a;
        private final ResolvedValue b;

        public MathFunction(String operator, ResolvedValue a, ResolvedValue b) {
            this.operator = operator;
            this.a = a;
            this.b = b;
        }

        @Override
        public String asText() {
            return "(" + asTextNoBrackets() + ")";
        }

        private String asTextNoBrackets() {
            return format(a)
                    + operator
                    + format(b);
        }

        private String format(ResolvedValue v) {
            if (v instanceof MathFunction mv) {
                return switch (operator) {
                    case "+", "-" -> switch (mv.operator) {
                        case "+", "-" -> mv.asTextNoBrackets();
                        default -> FormulaOptimizer.format(v);
                    };
                    case "*", "/" -> switch (mv.operator) {
                        case "*", "/" -> mv.asTextNoBrackets();
                        default -> FormulaOptimizer.format(v);
                    };
                    default -> FormulaOptimizer.format(v);
                };
            }
            return FormulaOptimizer.format(v);
        }
    }

    private static class AnyFunction extends AbstractOptimizedFunction {
        private final List<ResolvedValue> values;

        private AnyFunction(List<ResolvedValue> values) {
            this.values = new ArrayList<>();
            boolean hasFalse = false;
            for (ResolvedValue next : values) {
                if (next instanceof AnyFunction anyFn) {
                    this.values.addAll(0, anyFn.values);
                    continue;
                }
                if (next.equals(ResolvedValue.FALSE)) {
                    hasFalse = true;
                    continue;
                }
                if (next.equals(ResolvedValue.TRUE)) {
                    this.values.clear();
                    this.values.add(ResolvedValue.TRUE);
                    return;
                }
                this.values.add(0, next);
            }

            if (this.values.isEmpty()) {
                this.values.add(ResolvedValue.of(!hasFalse));
            }
        }

        @Override
        public String asText() {
            if (values.size() == 1) {
                return format(this.values.get(0));
            }
            return "any(" + values.stream()
                    .map(FormulaOptimizer::format)
                    .collect(Collectors.joining(",")) + ")";
        }
    }

    private static class AllFunction extends AbstractOptimizedFunction {
        private final List<ResolvedValue> values;

        private AllFunction(List<ResolvedValue> values) {
            this.values = new ArrayList<>();
            for (ResolvedValue next : values) {
                if (next instanceof AllFunction allFn) {
                    this.values.addAll(0, allFn.values);
                    continue;
                }
                if (next.equals(ResolvedValue.TRUE)) {
                    continue;
                }
                if (next.equals(ResolvedValue.FALSE)) {
                    this.values.clear();
                    this.values.add(ResolvedValue.FALSE);
                    return;
                }
                this.values.add(0, next);
            }

            if (this.values.isEmpty()) {
                this.values.add(ResolvedValue.TRUE);
            }
        }

        @Override
        public String asText() {
            if (values.size() == 1) {
                return format(this.values.get(0));
            }
            return "all(" + values.stream()
                    .map(FormulaOptimizer::format)
                    .collect(Collectors.joining(",")) + ")";
        }
    }


    private static abstract class AbstractOptimizedFunction extends ResolvedValue {

        private static final UnsupportedOperationException NOT_AVAILABLE_EXCEPTION = new UnsupportedOperationException(
                "Not available for optimization");

        @Override
        public int asNumber() {
            throw NOT_AVAILABLE_EXCEPTION;
        }

        @Override
        public double asDecimal() {
            throw NOT_AVAILABLE_EXCEPTION;
        }

        @Override
        public boolean asBoolean() {
            throw NOT_AVAILABLE_EXCEPTION;
        }
    }
}

File: formula-cs/Formula/Optimize/FormulaOptimizer.cs
using System.Text;
using Formula.ShuntingYard;

namespace Formula.Optimize;

public class FormulaOptimizer
{
    private static readonly ShuntingYardParser Parser = ShuntingYardParser.Create()
            .Operator("^", 4, Associativity.Right, OpFn2((a, b) => a + "^" + b))
            .Operator("*", 3, Associativity.Left, (a,b) => new MathFunction("*", a, b))
            .Operator("/", 3, Associativity.Left, (a,b) => new MathFunction("/", a, b))
            .Operator("+", 2, Associativity.Left, (a,b) => new MathFunction("+", a, b))
            .Operator("-", 2, Associativity.Left, (a,b) => new MathFunction("-", a, b))
            .Operator("!", 2, Associativity.Left, OpFn1((a) => "!" + a))
            .Operator("<", 3, Associativity.Left, OpFn2((a, b) => a + "<" + b))
            .Operator("<=", 3, Associativity.Left, OpFn2((a, b) => a + "<=" + b))
            .Operator(">", 3, Associativity.Left, OpFn2((a, b) => a + ">" + b))
            .Operator(">=", 3, Associativity.Left, OpFn2((a, b) => a + ">=" + b))
            .Operator("==", 3, Associativity.Left, OpFn2((a, b) => a + "==" + b))
            .Operator("!=", 3, Associativity.Left, OpFn2((a, b) => a + "!=" + b))
            .Operator("AND", 1, Associativity.Left, (a, b) => new AllFunction(new List<ResolvedValue>(new[] {b, a})))
            .Operator("OR", 1, Associativity.Left, (a, b) => new AnyFunction(new List<ResolvedValue>(new[] {b, a})))
            .Operator("d", 4, Associativity.Left, OpFn2((a, b) => a + "d" + b))
            .Term("true", () => ResolvedValue.Of("true"))
            .Term("false", () => ResolvedValue.Of("false"))
            .Term("null", () => ResolvedValue.Of("null"))
            .Function("abs", OpFn1(a => $"abs({a})"))
            .Function("min", OpFn2((a, b) => $"min({a},{b})"))
            .Function("max", OpFn2((a, b) => $"max({a},{b})"))
            .Function("floor", OpFn1(a => $"floor({a})"))
            .Function("ceil", OpFn1(a => $"ceil({a})"))
            .Function("signed", OpFn1(a => $"signed({a})"))
            .Function("if", OpFn3((a, b, c) => $"if({a},{b},{c})"))
            .Function("concat", OpFn2((a, b) => $"concat({a},{b})"))
            .Function("ordinal", OpFn1(a => $"ordinal({a})"))
            .Function("any", a => new AnyFunction(new List<ResolvedValue>(a)))
            .Function("all", a => new AllFunction(new List<ResolvedValue>(a)))
            .Variable("@", (_, key) => ResolvedValue.Of($"@{key}"))
            .Variable("@{", "}", (_, key) => ResolvedValue.Of($"@{{{key}}}"))
            .Variable("min(@", ")", (_, key) => ResolvedValue.Of($"min(@{key})"))
            .Variable("max(@", ")", (_, key) => ResolvedValue.Of($"max(@{key})"))
            .Variable("sum(@", ")", (_, key) => ResolvedValue.Of($"sum(@{key})"))
            .Comment("[", "]", (value, comment) => NamedResolvedValue.Of(value, comment.Substring(1, comment.Length - 2)))
            ;

    public static string Optimize(string formulaText) {
        var resolved = Parser.Parse(formulaText).Resolve();
        if (resolved is MathFunction mf) {
            return mf.AsTextNoBrackets();
        }
        return Format(resolved);
    }

    private static string Format(ResolvedValue value)
    {
        return value switch
        {
            QuotedTextResolvedValue quoted => quoted.AsQuotedText(),
            NamedResolvedValue named => value.AsText() + "[" + named.AsName() + "]",
            _ => value.AsText()
        };
    }

    private static Func<ResolvedValue, ResolvedValue> OpFn1(Func<string, string> fn) {
        return a => ResolvedValue.Of(fn.Invoke(Format(a)));
    }

    private static Func<ResolvedValue, ResolvedValue, ResolvedValue> OpFn2(Func<string, string, string> fn) {
        return (a, b) => ResolvedValue.Of(fn.Invoke(Format(a), Format(b)));
    }

    private static Func<ResolvedValue, ResolvedValue, ResolvedValue, ResolvedValue> OpFn3(Func<string, string, string, string> fn) {
        return (a, b, c) => ResolvedValue.Of(fn.Invoke(Format(a), Format(b), Format(c)));
    }

    private class MathFunction : AbstractOptimizedFunction {
        private readonly string _operator;
        private readonly ResolvedValue _a;
        private readonly ResolvedValue _b;

        internal MathFunction(string op, ResolvedValue a, ResolvedValue b)
        {
            _operator = op;
            _a = a;
            _b = b;
        }
        
        public override string AsText() {
            return "(" + AsTextNoBrackets() + ")";
        }

        public string AsTextNoBrackets() {
            return FormatMath(_a)
                   + _operator
                   + FormatMath(_b);
        }

        private string FormatMath(ResolvedValue v) 
        {
            if (v is MathFunction mv)
            {
                return _operator switch
                {
                    "+" or "-" => mv._operator switch
                    {
                        "+" or "-" => mv.AsTextNoBrackets(),
                        _ => Format(v)
                    },
                    "*" or "/" => mv._operator switch
                    {
                        "*" or "/" => mv.AsTextNoBrackets(),
                        _ => Format(v)
                    },
                    _ => Format(v)
                };
            }
            return Format(v);
        }
    }

    private class AnyFunction : AbstractOptimizedFunction {
        private readonly List<ResolvedValue> _values;

        internal AnyFunction(List<ResolvedValue> values) {
            _values = new List<ResolvedValue>();
            var hasFalse = false;
            foreach (var next in values) 
            {
                if (next is AnyFunction anyFn) {
                    _values.InsertRange(0, anyFn._values);
                    continue;
                }

                if (next.Equals(False))
                {
                    hasFalse = true;
                    continue;
                }

                if (next.Equals(True))
                {
                    _values.Clear();
                    _values.Add(True);
                    return;
                }
                _values.Insert(0, next);
            }

            if (_values.Count == 0)
            {
                _values.Add(hasFalse ? False : True);
            }
        }
        
        public override string AsText() {
            if (_values.Count == 1) {
                return Format(_values[0]);
            }

            var builder = new StringBuilder();
            builder.Append("any(");
            for (var i = 0; i < _values.Count; i++)
            {
                var next = _values[i];
                if (i > 0)
                {
                    builder.Append(',');
                }

                builder.Append(Format(next));
            }

            builder.Append(')');
            return builder.ToString();
        }
    }

    private class AllFunction : AbstractOptimizedFunction 
    {
        private readonly List<ResolvedValue> _values;

        internal AllFunction(List<ResolvedValue> values) {
            _values = new List<ResolvedValue>();
            foreach (var next in values) 
            {
                if (next is AllFunction allFn) {
                    _values.InsertRange(0, allFn._values);
                    continue;
                }

                if (next.Equals(True))
                {
                    continue;
                }

                if (next.Equals(False))
                {
                    _values.Clear();
                    _values.Add(False);
                    return;
                }
                _values.Insert(0, next);
            }

            if (_values.Count == 0)
            {
                _values.Add(True);
            }
        }
        
        public override string AsText() 
        {
            if (_values.Count == 1) {
                return Format(_values[0]);
            }

            var builder = new StringBuilder();
            builder.Append("all(");
            for (var i = 0; i < _values.Count; i++)
            {
                var next = _values[i];
                if (i > 0)
                {
                    builder.Append(',');
                }

                builder.Append(Format(next));
            }

            builder.Append(')');
            return builder.ToString();
        }
    }


    private abstract class AbstractOptimizedFunction : ResolvedValue 
    {
        private static readonly InvalidOperationException NotAvailableException = new("Not available for optimization");
        
        public override int AsNumber() {
            throw NotAvailableException;
        }
        
        public override double AsDecimal() {
            throw NotAvailableException;
        }
        
        public override bool AsBoolean() {
            throw NotAvailableException;
        }
    }
}